core/src/solution_submission/retry.rs

```diff
@@ -6,9 +6,60 @@ use ethcontract::{
     errors::{ExecutionError, MethodError},
     U256,
 };
+use futures::future::{BoxFuture, FutureExt as _};
 
 use super::MIN_GAS_PRICE_INCREASE_FACTOR;
 
+pub struct Args<'a> {
```

Comment:
I like this! That way arguments aren't positional 🎉

Reply:
Is there a reason for the non-dynamic per instance state (contract, gas_price_estimating) to be passed on every invocation? I can see how the gas_price_cap is already passed as we want to make it dependent on each solution going forward.

Also, I'm wondering if the abstraction should be built on to of a MethodBuilder instead of the concrete set of parameters for a solution submission. This way the function would just know how to retry some ethereum tx up to a  certain gas cap.

Reply:
This is a bigger task that I find hard to estimate how complex it will be so I don't want to undertake it now. In a future pr this logic is also going to change to fix the nonce errors. We can look it again after that.